debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/legacy/EventBuffer.java

```diff
@@ -0,0 +1,232 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.connector.mysql.legacy;
+
+import java.util.Queue;
+import java.util.concurrent.ArrayBlockingQueue;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.shyiko.mysql.binlog.event.Event;
+import com.github.shyiko.mysql.binlog.event.EventType;
+import com.github.shyiko.mysql.binlog.event.QueryEventData;
+
+import io.debezium.connector.mysql.legacy.BinlogReader.BinlogPosition;
+
+/**
+ * This class represents a look-ahead buffer that allows Debezium to accumulate binlog events and decide
+ * if the last event in transaction is either {@code ROLLBACK} or {@code COMMIT}. The incoming events are either
+ * supposed to be in transaction or out-of-transaction. When out-of-transaction they are sent directly into
+ * the destination handler. When in transaction the event goes through the buffering.
+ * <p>
+ * The reason for the buffering is that the binlog contains rolled back transactions in some cases. E.g. that's
+ * the case when a temporary table is dropped (see DBZ-390). For rolled back transactions we may not propagate
+ * any of the contained events, hence the buffering is applied.
+ * <p>
+ * The transaction start is identified by a {@code BEGIN} event. Transaction is ended either by {@code COMMIT}
+ * event or by {@code XID} an event.
+ * <p>
+ * If there are more events that can fit to the buffer then:
+ * <ul>
+ *     <li>Binlog position is recorded for the first event not fitting into the buffer</li>
+ *     <li>Binlog position is recorded for the commit event</li>
+ *     <li>Buffer content is sent to the final handler</li>
+ *     <li>Binlog position is rewound and all events between the above recorded positions are sent to the final handler</li>
+ * </ul>
+ *
+ * @author Jiri Pechanec
+ *
+ */
+class EventBuffer {
```

Comment:
Do we have this twice now?

Reply:
Yes, it was quite difficult to keep the class same for legacy and non-legacy classes due to class dependencies so I opted for making a copy.